config-system.lisp

```diff
@@ -0,0 +1,157 @@
+
+(define-condition invalid-datum-error (config-error)
+  ((place-symbol :initarg :place :initform nil
+		 :accessor invalid-datum-error-place)
+   (value :initarg :value :initform nil
+	  :accessor invalid-datum-error-value))
+  (:report
+   (lambda (c s)
+     (with-slots (place-symbol value) c
+       (format s "The value ~S is invalid for variable ~S." value place-symbol)))))
+
+(define-condition config-not-found-error (config-error)
+  ((place-symbol :initarg :place :initform nil
+                 :accessor config-not-found-error-place
+                 :type symbol)
+   (alternatives :initarg :alternatives :initform nil
+                 :accessor config-not-found-alternatives
+                 :type list))
+  (:report
+   (lambda (c s)
+     (with-slots (place-symbol alternatives) c
+       (if alternatives
+        (format s  "Setting ~A not found. Did you mean one of these? ~A"
+                (%full-symbol-string place-symbol) alternatives)
+        (format s  "Setting ~A not found." (%full-symbol-string place-symbol)))))))
+
+(defun %generate-config-var-code (declare-type name default validator documentation)
+  (check-type documentation string)
+  (with-gensyms (default-value is-valid func)
+    `(progn
+       (,declare-type ,name ,default ,documentation)
+       (let* ((,default-value ,name)
+              (,func ,validator)
+              (,is-valid (funcall ,func ,default-value)))
+         (if ,is-valid
+             (setf (gethash (quote ,name) *config-vars*)
+                   (make-config-info :name (quote ,name) :default ,default-value
+                                     :validator ,func :doc ,documentation))
+             (error 'invalid-datum-error :place (quote ,name) :value ,default-value))))))
+
+(defmacro define-config-parameter (name default validator documentation)
+  "Create and register a configurable variable with the given default value,
+validator function, and documentation using DEFPARAMETER"
+  ;; if we are given a lambda, we need to evaluate it in order to call it
+  (%generate-config-var-code 'defparameter name default validator documentation))
+
+(defmacro define-config-var (name default validator documentation)
+  "Create and register a configurable variable with the given default value,
+validator function, and documentation using DEFVAR"
+  ;; if we are given a lambda, we need to evaluate it in order to call it
+  (%generate-config-var-code 'defvar name default validator documentation))
+
+(defmacro define-config-enum (name default values documentation)
+  "Same as DEFINE-CONFIG-PARAMETER, except the validation function
+is built by creating a function that checks if the set  value is in the VALUES list."
+  ;; if we wanted to get fancy, we could define a special type for this enumeration at this point too.
+  `(define-config-parameter ,name ,default (lambda (x) (member x ,values :test #'equal))
+                            ,documentation))
+
+(defun list-all-configurations ()
```

Comment:
I'm not a big fan of this function name. Maybe `get-all-config-info`?

Reply:
What has been recommended to me (i think by puercopop) was to avoid using `get-` in getters. I think a name like `all-config-info` would be better. As mentioned in another response, this function could also benefit from a database optional or key variable, if we go that route.